From 259dff25a78a23ee35dfc9a7ae20ca1013b1f9fa Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 3 Jul 2014 18:42:37 +0200
Subject: [PATCH] app: report disconnection error details

It may be useful to provide more detailed reason for disconnection.

Resolves: rhbz#1115986
(cherry picked from commit 225dff3cb068d8920d647d1e0a07f3f352bbc8d4)
---
 src/virt-viewer-app.c           | 6 ++++--
 src/virt-viewer-session-spice.c | 6 +++---
 src/virt-viewer-session-vnc.c   | 2 +-
 src/virt-viewer-session.c       | 5 +++--
 src/virt-viewer-session.h       | 2 +-
 5 files changed, 12 insertions(+), 9 deletions(-)

diff --git a/src/virt-viewer-app.c b/src/virt-viewer-app.c
index 0110a9f..b05e966 100644
--- a/src/virt-viewer-app.c
+++ b/src/virt-viewer-app.c
@@ -75,6 +75,7 @@ static void virt_viewer_app_connected(VirtViewerSession *session,
 static void virt_viewer_app_initialized(VirtViewerSession *session,
                                         VirtViewerApp *self);
 static void virt_viewer_app_disconnected(VirtViewerSession *session,
+                                         const gchar *msg,
                                          VirtViewerApp *self);
 static void virt_viewer_app_auth_refused(VirtViewerSession *session,
                                          const char *msg,
@@ -1394,7 +1395,7 @@ virt_viewer_app_initialized(VirtViewerSession *session G_GNUC_UNUSED,
 }
 
 static void
-virt_viewer_app_disconnected(VirtViewerSession *session G_GNUC_UNUSED,
+virt_viewer_app_disconnected(VirtViewerSession *session G_GNUC_UNUSED, const gchar *msg,
                              VirtViewerApp *self)
 {
     VirtViewerAppPrivate *priv = self->priv;
@@ -1410,6 +1411,7 @@ virt_viewer_app_disconnected(VirtViewerSession *session G_GNUC_UNUSED,
         GtkWidget *dialog = virt_viewer_app_make_message_dialog(self,
             _("Unable to connect to the graphic server %s"), priv->pretty_address);
 
+        g_object_set(dialog, "secondary-text", msg, NULL);
         gtk_dialog_run(GTK_DIALOG(dialog));
         gtk_widget_destroy(dialog);
     }
@@ -1422,7 +1424,7 @@ static void virt_viewer_app_cancelled(VirtViewerSession *session,
 {
     VirtViewerAppPrivate *priv = self->priv;
     priv->cancelled = TRUE;
-    virt_viewer_app_disconnected(session, self);
+    virt_viewer_app_disconnected(session, NULL, self);
 }
 
 
diff --git a/src/virt-viewer-session-spice.c b/src/virt-viewer-session-spice.c
index 097c9d7..ad3028a 100644
--- a/src/virt-viewer-session-spice.c
+++ b/src/virt-viewer-session-spice.c
@@ -519,12 +519,12 @@ virt_viewer_session_spice_main_channel_event(SpiceChannel *channel G_GNUC_UNUSED
         break;
     case SPICE_CHANNEL_ERROR_CONNECT:
         g_debug("main channel: failed to connect");
-        g_signal_emit_by_name(session, "session-disconnected");
+        g_signal_emit_by_name(session, "session-disconnected", NULL);
         break;
     case SPICE_CHANNEL_ERROR_IO:
     case SPICE_CHANNEL_ERROR_LINK:
     case SPICE_CHANNEL_ERROR_TLS:
-        g_signal_emit_by_name(session, "session-disconnected");
+        g_signal_emit_by_name(session, "session-disconnected", NULL);
         break;
     default:
         g_warning("unhandled spice main channel event: %d", event);
@@ -815,7 +815,7 @@ virt_viewer_session_spice_channel_destroy(G_GNUC_UNUSED SpiceSession *s,
 
     self->priv->channel_count--;
     if (self->priv->channel_count == 0)
-        g_signal_emit_by_name(self, "session-disconnected");
+        g_signal_emit_by_name(self, "session-disconnected", NULL);
 }
 
 #define UUID_LEN 16
diff --git a/src/virt-viewer-session-vnc.c b/src/virt-viewer-session-vnc.c
index 545caaa..e4a6c93 100644
--- a/src/virt-viewer-session-vnc.c
+++ b/src/virt-viewer-session-vnc.c
@@ -115,7 +115,7 @@ virt_viewer_session_vnc_disconnected(VncDisplay *vnc G_GNUC_UNUSED,
     virt_viewer_session_clear_displays(VIRT_VIEWER_SESSION(session));
     display = virt_viewer_display_vnc_new(session->priv->vnc);
     g_debug("Disconnected");
-    g_signal_emit_by_name(session, "session-disconnected");
+    g_signal_emit_by_name(session, "session-disconnected", NULL);
     virt_viewer_display_set_show_hint(VIRT_VIEWER_DISPLAY(display),
                                       VIRT_VIEWER_DISPLAY_SHOW_HINT_READY, FALSE);
 }
diff --git a/src/virt-viewer-session.c b/src/virt-viewer-session.c
index 20d5fb1..9410004 100644
--- a/src/virt-viewer-session.c
+++ b/src/virt-viewer-session.c
@@ -220,9 +220,10 @@ virt_viewer_session_class_init(VirtViewerSessionClass *class)
                  G_SIGNAL_RUN_FIRST,
                  G_STRUCT_OFFSET(VirtViewerSessionClass, session_disconnected),
                  NULL, NULL,
-                 g_cclosure_marshal_VOID__VOID,
+                 g_cclosure_marshal_VOID__STRING,
                  G_TYPE_NONE,
-                 0);
+                 1,
+                 G_TYPE_STRING);
 
     g_signal_new("session-channel-open",
                  G_OBJECT_CLASS_TYPE(object_class),
diff --git a/src/virt-viewer-session.h b/src/virt-viewer-session.h
index 388d675..316fdf0 100644
--- a/src/virt-viewer-session.h
+++ b/src/virt-viewer-session.h
@@ -78,7 +78,7 @@ struct _VirtViewerSessionClass {
     /* signals */
     void (*session_connected)(VirtViewerSession *session);
     void (*session_initialized)(VirtViewerSession *session);
-    void (*session_disconnected)(VirtViewerSession *session);
+    void (*session_disconnected)(VirtViewerSession *session, const gchar *msg);
     void (*session_auth_refused)(VirtViewerSession *session, const gchar *msg);
     void (*session_auth_failed)(VirtViewerSession *session, const gchar *msg);
     void (*session_usb_failed)(VirtViewerSession *session, const gchar *msg);
